Enhancement: Adds position averaging transformation

Lp0lp opened the pull request:
Changes made in this Pull Request:
 - Adds a position averaging transformation that returns the atom coordinates averaged across the previous N frames;
 - Makes use of the transformations API.

This works by storing the coordinates of the previous N frames in a class attribute and replacing the timestep coordinates by the calculated mean. While iterating frames < N it calculates the average positions with the frames iterated up to that point. This can be followed by comparing the `current_avg` attribute to `avg_frames`.
It contains a built in override `check_reset`, which will reset the averaging if non sequential iterations are detected, or if the iteration finishes. This behavior can be disabled if `check_reset=False`. Manual resets can be performed by the user by calling the `.resetarrays()` method.

As of right now no Tests have yet been prepared as I was hoping to hear some feedback/suggestions first!



PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

richardjgowers commented:
The code looks solid enough, so next up would be (at least) regression
tests. You could also build a Universe on top of a numpy array and have
nice analytical expected results. Out of interest, what’s the application
for this?
On Thu, Feb 21, 2019 at 2:14 PM, codecov[bot] <notifications@github.com>
wrote:

> Codecov
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=h1>
> Report
>
> Merging #2208
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=desc>
> into develop
> <https://codecov.io/gh/MDAnalysis/mdanalysis/commit/839410ad63154503c87237ed08dc5a2ec7c1a8c2?src=pr&el=desc>
> will *decrease* coverage by 1.44%.
> The diff coverage is 27.5%.
>
> [image: Impacted file tree graph]
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=tree>
>
> @@             Coverage Diff             @@
> ##           develop    #2208      +/-   ##
>
> ===========================================
> - Coverage    90.71%   89.27%   -1.45%
>
> ===========================================
>
>   Files           15      159     +144
>
>   Lines         1982    19722   +17740
>
>   Branches         0     2772    +2772
>
> ===========================================
> + Hits          1798    17607   +15809
> - Misses         184     1512    +1328
> - Partials         0      603     +603
>
> Impacted Files
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=tree> Coverage
> Δ
> package/MDAnalysis/transformations/__init__.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL3RyYW5zZm9ybWF0aW9ucy9fX2luaXRfXy5weQ==> 100%
> <100%> (ø)
> ...ge/MDAnalysis/transformations/positionaveraging.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL3RyYW5zZm9ybWF0aW9ucy9wb3NpdGlvbmF2ZXJhZ2luZy5weQ==> 25.64%
> <25.64%> (ø)
> package/MDAnalysis/coordinates/GRO.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL2Nvb3JkaW5hdGVzL0dSTy5weQ==> 93.87%
> <0%> (ø)
> package/MDAnalysis/coordinates/base.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL2Nvb3JkaW5hdGVzL2Jhc2UucHk=> 93.3%
> <0%> (ø)
> package/MDAnalysis/analysis/helanal.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL2FuYWx5c2lzL2hlbGFuYWwucHk=> 85.19%
> <0%> (ø)
> package/MDAnalysis/topology/MinimalParser.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL3RvcG9sb2d5L01pbmltYWxQYXJzZXIucHk=> 100%
> <0%> (ø)
> package/MDAnalysis/visualization/__init__.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL3Zpc3VhbGl6YXRpb24vX19pbml0X18ucHk=> 100%
> <0%> (ø)
> package/MDAnalysis/core/qcprot.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL2NvcmUvcWNwcm90LnB5> 100%
> <0%> (ø)
> ...e/MDAnalysis/analysis/encore/clustering/cluster.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL2FuYWx5c2lzL2VuY29yZS9jbHVzdGVyaW5nL2NsdXN0ZXIucHk=> 96.22%
> <0%> (ø)
> package/MDAnalysis/coordinates/DCD.py
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree#diff-cGFja2FnZS9NREFuYWx5c2lzL2Nvb3JkaW5hdGVzL0RDRC5weQ==> 97.16%
> <0%> (ø)
> ... and 136 more
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208/diff?src=pr&el=tree-more>
> ------------------------------
>
> Continue to review full report at Codecov
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=continue>
> .
>
> *Legend* - Click here to learn more
> <https://docs.codecov.io/docs/codecov-delta>
> Δ = absolute <relative> (impact), ø = not affected, ? = missing data
> Powered by Codecov
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=footer>.
> Last update 839410a...a6301ee
> <https://codecov.io/gh/MDAnalysis/mdanalysis/pull/2208?src=pr&el=lastupdated>.
> Read the comment docs <https://docs.codecov.io/docs/pull-request-comments>
> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/2208#issuecomment-466149381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jBz6wA6fzDo12gdDxTBCHNoiI5aOSks5vPv4sgaJpZM4bIMT0>
> .
>

Lp0lp commented:
> The code looks solid enough, so next up would be (at least) regression tests. You could also build a Universe on top of a numpy array and have nice analytical expected results.

Thank you, will get on with the tests!

> Out of interest, what’s the application for this?

This was done to more easily identify the formation of the acyl-chain hexagonal matrix that is associated with the formation of lipid gel phases in membrane bilayers, but I believe it could become useful for other calculations!

Lp0lp commented:
Have just added a few tests to the testing package. They test the results of the averaging whilst iterating forwards and/or backwards through the trajectory. Also added a test to make sure the reset is working as intended. 
Additionally, I found out that the __init__.py file for the transformations tests was missing, and therefore the tests were not being ran by pytest. I have fixed this here but should I open a separate bug report?

orbeckst commented:
> Additionally, I found out that the __init__.py file for the transformations tests was missing, and therefore the tests were not being ran by pytest. I have fixed this here but should I open a separate bug report?

Yes, please – that one is easier to review and will likely get merged quickly. Thanks!

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -0,0 +1,102 @@
+
+
+"""
+Trajectory Coordinate Averaging --- 
```
remove `---`

orbeckst replied:
make sure that the docs are integrated into the docs under [doc/sphinx/source/documentation_pages/transformations](https://github.com/MDAnalysis/mdanalysis/tree/develop/package/doc/sphinx/source/documentation_pages/transformations)

Please build the docs locally and check, see https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#generating-the-docs

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -0,0 +1,102 @@
+
+
+
+class PositionAverager:  
```
For Python 2.7: should inherit from `object`

orbeckst commented on testsuite/MDAnalysisTests/transformations/test_positionaveraging.py:
```diff
@@ -0,0 +1,66 @@
+    for ts in posaveraging_universes.trajectory:
+        avgd[...,ts.frame] = ts.positions.copy()
+        
+    assert_array_almost_equal(ref_matrix_fwd, avgd[1,:,-1], decimal = 5)   
```
PEP8: no space around `=` in argument lists

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -0,0 +1,102 @@
+    
+    Example
+    -------
+    
```
The example needs to be in a code block. Also add some text to explain what happens and what the parameters mean. 

How do you use it? What could you meaningfully compute in the example? People want to see what this can do for them and the best way to show it is by example (or examples).

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -0,0 +1,102 @@
+    Parameters
+    ----------
+    avg_frames: int
+    Determines the number of frames to be used for the position averaging. 
```
indent

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -0,0 +1,102 @@
+    
+    
+    check_reset: bool, optional
+    If 'True', position averaging will be reset, and a warning raised,
```
indent

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -0,0 +1,102 @@
+    
+    Averages the coordinates of a given timestep with the N previous frames. 
+    For frames < N, the average of the frames iterated up to that point will be
+    returned.  
```
Explain what happens when
- direction is reversed
- frames are skipped
- arbitrary frames are accessed

orbeckst commented on testsuite/MDAnalysisTests/transformations/test_positionaveraging.py:
```diff
@@ -0,0 +1,66 @@
+import MDAnalysis as md
+from MDAnalysis.transformations import PositionAverager
+from MDAnalysisTests import datafiles
+
```
add a test for random access of frames (eg frames 2, 5, 1, 2, 2, 8)

orbeckst commented:
Hi @Lp0lp , sorry it took as while to review your PR. Are you still interested in finishing it? It doesn't look as if there's a large number of things to do.

orbeckst commented:
Can you please resolve the conflicts?

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
     Example
     -------
+
+    e.g. Average the coordinates of a given AtomGroup over the course of
```
You can leave out "e.g." and just start the sentence.

orbeckst replied:
Also applies below.

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
+    e.g. Average the coordinates of a given AtomGroup over the course of
+    the previous N frames. For N=3, the output will correspond to the
+    average of the coordinates over the last 3 frames.
+    When check_reset=True, the averager will be reset once the iteration is
```
When talking about code in text, put it into double backtics
```reST
When ``check_reset=True``, the averager will be 
```

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
+            print(ts.positions)
     
-    u.trajectory.add_transformations(PositionAverager(N, check_reset=True))    
+    e.g. When check_reset=False, the average of coordinates from non
```
formatting of code in text

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
     
+    e.g. For frames < N, the average is calculated with the frames iterated up
+    to that point and thus will not follow the same behaviour as for the
+    frames > N. This behaviour can be followed by comparing current_avg to 
```
Write ``current_avg`` and all other variable names with double-backtics.

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
+        frames = [0, 7, 1, 6]        
+        transformation.resetarrays()
+        for ts in u.trajectory[frames]:
+            print(ts.positions)
```
Briefly explain what you expect to see when this code runs. I read the docs and it is not evident to me what the example code will do. Make it easy for the reader, don't expect them to deduce what should happen. Tell them. This is important when someone tries out the code. They want to have reassurance that they get the right thing when they try it for themselves.

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
+        u.trajectory.add_transformations(transformation) 
+        for ts in u.trajectory:
+            if transformation.current_avg == transformation.avg_frames:
+                print(ts.positions)
```
Please explain what this code is doing. See comment above.

orbeckst commented on package/MDAnalysis/transformations/positionaveraging.py:
```diff
@@ -20,34 +39,73 @@
-    If 'True', position averaging will be reset, and a warning raised,
-    if the trajectory iteration direction changes. If 'False', position
-    averaging will not reset, regardless of the iteration.
+        If 'True', position averaging will be reset, and a warning raised,
```
Use single backtics when referencing kwargs and double backtics for code and ``True`` and ``False`` so that they are typeset consistently.

Lp0lp commented:
> Looking good, just some final requests for the docs.

Thank you for the very good feedback! Have fixed the syntax mistakes that were pointed out. Have also described all the examples much more in depth. They should be more easily understood now, but I am sure there is more room for improvement. Should I include more/different examples? or phrase them differently?

orbeckst commented:
@Lp0lp thanks for your contribution!

orbeckst commented:
@Lp0lp I invited you to become a member of the MDAnalysis org and of the MDAnalysis contributors team.
